### PR TITLE
Add spec and parser for dmsetup_status

### PR DESCRIPTION
**Pull request by JoySnow**
Signed-off-by: XiaoXue Wang <xiaoxwan@redhat.com>

### CODE REVIEW

File: insights/parsers/dmsetup.py [hunk 1]
```diff
@@ -121,3 +130,155 @@ def __getitem__(self, idx):
...
+        """
+        Format:
+            <transaction_id> <used_metadata_blocks>/<total_metadata_blocks>
+            <used_data_blocks>/<total_data_blocks> <held_metadata_root>
+            ro|rw|out_of_data_space [no_]discard_passdown [error|queue]_if_no_space
+            needs_check|- metadata_low_watermark
+        Refer to https://www.kernel.org/doc/Documentation/device-mapper/thin-provisioning.txt .
+        """
+        args = target_args.split()
+        if len(args) < 8 or '/' not in args[1] or '/' not in args[2]:
```
Comment (xiangce) [inline insights/parsers/dmsetup.py:228]: The same as above: here just skip it w/o any prompt info?

File: insights/parsers/dmsetup.py [hunk 2]
```diff
@@ -121,3 +130,155 @@ def __getitem__(self, idx):
...
+        return parsed_args
+
+    def _parse_target_args_thin(self, target_args):
+        """
+        Format:
+            <nr_mapped_sectors> <highest_mapped_sector>
+        Refer to https://www.kernel.org/doc/Documentation/device-mapper/thin-provisioning.txt .
+        """
+        args = target_args.split()
+        if len(args) < 2:
```
Comment (xiangce) [inline insights/parsers/dmsetup.py:246]: The same as above: here just skip it w/o any prompt info?

File: insights/parsers/dmsetup.py [hunk 3]
```diff
@@ -121,3 +130,155 @@ def __getitem__(self, idx):
...
+        return parsed_args
+
+    def _parse_target_args_snapshot(self, target_args):
+        """
+        Format:
+            <sectors_allocated>/<total_sectors> <metadata_sectors>
+        Refer to https://www.kernel.org/doc/Documentation/device-mapper/snapshot.txt .
+        """
+        args = target_args.split()
+        if len(args) < 2 or '/' not in args[0]:
```
Comment (xiangce) [inline insights/parsers/dmsetup.py:260]: The same as above: here just skip it w/o any prompt info?

**Review by xiangce — CHANGES_REQUESTED** [review]

File: insights/parsers/dmsetup.py [hunk 4]
```diff
@@ -9,6 +9,9 @@
 DmsetupInfo - command ``dmsetup info -C``
 -----------------------------------------
 
+DmsetupStatus - command ``dmsetup status``
+------------------------------------------
```
Comment (xiangce) [inline insights/parsers/dmsetup.py:13]: This is not good adding 2 titles for 1 module, please just leave one which is general for this module and make the current 2 as sub titles.

File: insights/parsers/dmsetup.py [hunk 5]
```diff
@@ -121,3 +130,155 @@ def __getitem__(self, idx):
...
+            self.data.append(SetupStatus(
+                device_name=device_name,
+                start_sector=device_info_spl[0],
+                num_sectors=device_info_spl[1],
+                target_type=target_type,
+                target_args=target_args,
+                parsed_args=parsed_args,
+            ))
+        self.names = [dm[0] for dm in self.data]
+        self.by_name = dict((dm[0], dm) for dm in self.data)
```
[the remark this answers is not in the document]
Comment (JoySnow) [reply insights/parsers/dmsetup.py:211]: Sure, I changed to use `@property` for `self.names` and `self.by_name`.

File: insights/parsers/dmsetup.py [hunk 6]
```diff
@@ -121,3 +130,155 @@ def __getitem__(self, idx):
...
+        data = [
+            SetupStatus(
+                device_name='rootvg-tanlv', start_sector='0',
+                num_sectors='6291456', target_type='linear',
+                target_args=None, parsed_args=None,
+            ), ...
+        ]
+
+    Attributes:
+        data (list): List of devices found, in order using SetupStatus structure
```
[the remark this answers is not in the document]
Comment (JoySnow) [reply insights/parsers/dmsetup.py:166]: Change to use `list` as a base class.

File: insights/parsers/dmsetup.py [hunk 7]
```diff
@@ -121,3 +130,155 @@ def __getitem__(self, idx):
...
+        """
+        args = target_args.split()
+        if len(args) < 2 or '/' not in args[0]:
+            return
+        parsed_args = {}
+        parsed_args['sectors_allocated'], parsed_args['total_sectors'] = args[0].split('/', 1)
+        parsed_args['metadata_sectors'] = args[1]
+        return parsed_args
+
+    def __len__(self):
```
[the remark this answers is not in the document]
Comment (JoySnow) [reply insights/parsers/dmsetup.py:267]: Yep.

File: insights/parsers/dmsetup.py [hunk 8]
```diff
@@ -121,3 +130,155 @@ def __getitem__(self, idx):
...
+        ['rw', 'no_discard_passdown', 'queue_if_no_space', '-']
+    """
+
+    def parse_content(self, content):
+        self.data = []
+        for line in content:
+            _device_name, _device_info_str = line.rsplit(':', 1)
+            device_name = _device_name.strip()
+            device_info_spl = _device_info_str.strip().split(' ', 3)
+            if len(device_info_spl) < 3:
```
[the remark this answers is not in the document]
Comment (JoySnow) [reply insights/parsers/dmsetup.py:195]: Changed to capture such lines into `self.unparseable_lines`.

File: insights/parsers/dmsetup.py (hunk 1 above)
Comment (JoySnow) [reply insights/parsers/dmsetup.py:228]: Changed to capture such lines into `self.unparseable_lines`.

File: insights/parsers/dmsetup.py (hunk 3 above)
Comment (JoySnow) [reply insights/parsers/dmsetup.py:260]: Changed to capture such lines into `self.unparseable_lines`.

File: insights/parsers/dmsetup.py (hunk 4 above)
Comment (xiangce) [reply insights/parsers/dmsetup.py:13]: My bad, I forgot to unfold the closet code above

**Comment by JoySnow** [comment]
Hi @xiangce, I had updated according to your suggestions. Could you help re-review this MR?

**Review by xiangce — APPROVED** [review]

**Review by xiangce — APPROVED** [review]

**Pull request merged**